src/merkle/merkle-objects.ts

```diff
@@ -1,49 +1,187 @@
+   */
+  public get failedRequests(): Request[] {
+    return this._failedRequests
+  }
+
+  /**
+   * All requests that were rejected by Ceramic's conflict resolution. Only available after
+   * calling 'setTipToAnchor'.
+   */
+  public get rejectedRequests(): Request[] {
+    return this._rejectedRequests
+  }
+
+  /**
+   * The Anchor database has a 1-1 foreign key reference to a single Request. This is a remnant
+   * of a time when every Anchor always directly corresponded to a single Request.  Now it's
+   * possible for one Anchor to satisfy multiple Requests on the same Stream, but the database
+   * still expects us to provide it a single Request to link. So we somewhat arbitrarily provide
+   * it the Request whose CID is latest in the log of all the Requests that were successfully
+   * anchored for this stream.
+   */
+  public get newestAcceptedRequest(): Request {
+    return this._newestAcceptedRequest
+  }
+
+  /**
+   * Returns true if this Stream was already anchored at the time that it was loaded during the
+   * anchoring process (most likely by another anchoring service after the creation of the original
+   * Request(s)).
+   */
+  public get alreadyAnchored(): boolean {
+    return this._alreadyAnchored
+  }
+
+  /**
+   * Marks that the CommitID corresponding to this Request could not be loaded from the Ceramic node.
+   * @param request
+   */
+  failRequest(request: Request): void {
+    this._failedRequests.push(request)
+  }
+
+  /**
+   * Marks that this Stream could not be loaded from the Ceramic node and we should therefore
+   * fail all pending requests on this Stream.
+   */
+  failAllRequests(): void {
+    this._failedRequests = this._requests
+  }
+
+  allRequestsFailed(): boolean {
+    return this._failedRequests.length == this._requests.length
+  }
+
+  shouldAnchor(): boolean {
+    return this.cid != null && this._acceptedRequests.length > 0
+  }
+
+  /**
+   * Given the current version of the stream, updates this.cid to include the appropriate tip to
+   * anchor.  Note that the CID selected may be the cid corresponding to any of the pending anchor
+   * requests, or to none of them if a newer, better CID is learned about from the Ceramic node.
+   * Also updates the Candidate's internal bookkeeping to keep track of which Requests
+   * were included in the tip being anchored and which were rejected by the Ceramic node's conflict
+   * resolution.
+   * @param stream
+   */
+  setTipToAnchor(stream: Stream): void {
+    if (stream.state.anchorStatus == AnchorStatus.ANCHORED) {
+      this._alreadyAnchored = true
+    } else {
+      this._cid = stream.tip
+      this._metadata = stream.metadata
+    }
+
+    // Check the log of the Stream that was loaded from Ceramic to see which of the pending requests
+    // are for CIDs that are included in the current version of the Stream's log.
+    const includedRequests = this._requests.filter((req) => {
```

Comment:
Might be able to reduce the number of loops and nesting here. Maybe not necessary right now but performance will continue to degrade per stream processed as the size of the stream log grows.

Reply:
I mean the `filter`/`includes`/`find` calls between line 146 and line 166.

Reply:
I think a stream log would be considered "large" if it had around a thousand entries in it, which is still going to be nothing for these loops.  So I'm not really worried about the perf cost here.  Especially since anchoring happens twice a day currently, this code is far from performance sensitive.  I'd rather avoid premature optimization until/unless we see an issue here